File: backend/go/internal/protocol/replay.go
package protocol

// ReplayWindow is an RFC 6479 style sliding anti-replay window of 1024 counters.
// Not safe for concurrent use; callers hold the session lock.
type ReplayWindow struct {
	top    uint64
	bitmap [WindowSize / 64]uint64
}

const WindowSize = 1024

// Check reports whether counter c is acceptable without modifying the window.
func (w *ReplayWindow) Check(c uint64) bool {
	if c == 0 {
		return false
	}
	if c > w.top {
		return true
	}
	if w.top-c >= WindowSize {
		return false
	}
	idx := c % WindowSize
	return w.bitmap[idx/64]&(1<<(idx%64)) == 0
}

// Update marks c as seen. It must only be called after Check(c) returned true
// and the datagram authenticated. It reports whether c advanced the window top.
func (w *ReplayWindow) Update(c uint64) (newTop bool) {
	if c > w.top {
		diff := c - w.top
		if diff >= WindowSize {
			w.bitmap = [WindowSize / 64]uint64{}
		} else {
			for i := w.top + 1; i <= c; i++ {
				idx := i % WindowSize
				w.bitmap[idx/64] &^= 1 << (idx % 64)
			}
		}
		w.top = c
		newTop = true
	}
	idx := c % WindowSize
	w.bitmap[idx/64] |= 1 << (idx % 64)
	return newTop
}

// CheckAndUpdate is Check followed by Update.
func (w *ReplayWindow) CheckAndUpdate(c uint64) bool {
	if !w.Check(c) {
		return false
	}
	w.Update(c)
	return true
}

File: backend/rust/src/protocol/replay.rs
//! RFC 6479 style anti-replay window (spec section 7.5).

pub const WINDOW_SIZE: u64 = 1024;
const WORDS: usize = (WINDOW_SIZE / 64) as usize;

#[derive(Debug, Clone, Default)]
pub struct ReplayWindow {
    top: u64,
    bitmap: [u64; WORDS],
}

impl ReplayWindow {
    fn bit(c: u64) -> (usize, u64) {
        let idx = c % WINDOW_SIZE;
        ((idx / 64) as usize, 1u64 << (idx % 64))
    }

    /// Whether counter `c` would be accepted. Does not modify the window.
    pub fn check(&self, c: u64) -> bool {
        if c == 0 {
            return false;
        }
        if c > self.top {
            return true;
        }
        if self.top - c >= WINDOW_SIZE {
            return false;
        }
        let (w, m) = Self::bit(c);
        self.bitmap[w] & m == 0
    }

    /// Mark `c` as seen (only after `check` and successful authentication).
    /// Returns true if `c` advanced the window.
    pub fn update(&mut self, c: u64) -> bool {
        let mut new_top = false;
        if c > self.top {
            if c - self.top >= WINDOW_SIZE {
                self.bitmap = [0; WORDS];
            } else {
                for i in self.top + 1..=c {
                    let (w, m) = Self::bit(i);
                    self.bitmap[w] &= !m;
                }
            }
            self.top = c;
            new_top = true;
        }
        let (w, m) = Self::bit(c);
        self.bitmap[w] |= m;
        new_top
    }

    pub fn check_and_update(&mut self, c: u64) -> bool {
        if !self.check(c) {
            return false;
        }
        self.update(c);
        true
    }
}
